### PR TITLE
Reduce number of concurrent open file handles

### PR DESCRIPTION
**Pull request by wbinnssmith**
This:

* Uses the graceful-fs package in place of the built-in fs module in `NodeFS`. graceful-fs queues calls and retries on EMFILE.
* Throttles the number of concurrent `asset.getCode()` calls in the HMRServer by using a PromiseQueue with a maximum concurrency limit, currently limited to 64 (macOS currently has a fairly low per-process fd limit of 256). `asset.getCode()` internally uses `fs.createReadStream()` through `Cache.getStream()`. 
  * **Note**: In my testing, graceful-fs was not sufficient to address excessive file handles and this queue in the HMRServer was necessary, as Node `require`'s own file reading threw an EMFILE.
  * **Question**: Should this happen in the Cache or even NodeFS instead?

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.05s | +7.00ms |
| Cached | 2.98s | +42.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 310.00ms | -33.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 253.00ms | +27.00ms ⚠️ |
| dist/legacy/index.45cccf70.js | 1.10kb | +0.00b | 2.51s | +229.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 2.61s | +213.00ms ⚠️ |
| dist/legacy/index.b3acc8eb.css | 77.00b | +0.00b | 2.50s | +226.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 195.00ms | -52.00ms 🚀 |
| dist/modern/index.1964d1f6.js | 1.10kb | +0.00b | 306.00ms | -31.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 373.00ms | -19.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 306.00ms | -31.00ms 🚀 |
| dist/modern/index.7bc22a8b.css | 77.00b | +0.00b | 306.00ms | -31.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 24.85s | -1.18s |
| Cached | 2.97s | -30.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/NotFound.f8a0616d.js | 387.00b | +0.00b | 859.00ms | -164.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 95.00ms | -178.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 478.81kb | +0.00b | 231.00ms | +13.00ms ⚠️ |
| dist/PermalinkedComment.9c5ad065.js | 4.22kb | +0.00b | 231.00ms | +14.00ms ⚠️ |
| dist/UserProfile.c1927888.js | 1.62kb | +0.00b | 231.00ms | +14.00ms ⚠️ |
| dist/NotFound.f8a0616d.js | 387.00b | +0.00b | 218.00ms | +54.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 130.00ms | -45.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.71m | -3.69s |
| Cached | 6.89s | -399.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/pdfRenderer.15a6a023.js | 1.11mb | +0.00b | 49.20s | +2.50s ⚠️ |
| dist/popup.6c453e3c.js | 169.96kb | +0.00b | 48.97s | +2.53s ⚠️ |
| dist/media-viewer.7079364e.js | 72.03kb | +0.00b | 14.17s | +739.00ms ⚠️ |
| dist/card.383bd9bd.js | 5.61kb | +0.00b | 14.17s | +736.00ms ⚠️ |
| dist/media-viewer.59c17dad.js | 3.92kb | +0.00b | 14.17s | +738.00ms ⚠️ |
| dist/dropzone.3c67f0a6.js | 3.23kb | +0.00b | 27.69s | +1.43s ⚠️ |
| dist/16.290b4a43.js | 2.38kb | +0.00b | 1.61s | +269.00ms ⚠️ |
| dist/card.eea6ad8c.js | 2.03kb | +0.00b | 27.68s | +1.42s ⚠️ |
| dist/date.231500d2.js | 1.73kb | +0.00b | 6.03s | +603.00ms ⚠️ |
| dist/16.d16f14ed.js | 1.73kb | +0.00b | 5.53s | +406.00ms ⚠️ |
| dist/images.80c92f5e.js | 1.67kb | +0.00b | 6.18s | +642.00ms ⚠️ |
| dist/workerHasher.6d5b22d7.js | 1.54kb | +0.00b | 1.47m | +1.28m ⚠️ |
| dist/list-number.c35f9589.js | 1.46kb | +0.00b | 6.18s | +642.00ms ⚠️ |
| dist/status.8ab4226b.js | 1.46kb | +0.00b | 6.43s | +704.00ms ⚠️ |
| dist/code.3dd71280.js | 1.38kb | +0.00b | 6.03s | +603.00ms ⚠️ |
| dist/16.51049e2c.js | 1.32kb | +0.00b | 1.61s | +269.00ms ⚠️ |
| dist/16.18344f47.js | 1.31kb | +0.00b | 5.81s | +507.00ms ⚠️ |
| dist/link.2f4427e6.js | 1.30kb | +0.00b | 6.18s | +642.00ms ⚠️ |
| dist/heading6.fdbbdf4d.js | 1.30kb | +0.00b | 6.55s | +744.00ms ⚠️ |
| dist/heading3.2113b7a1.js | 1.29kb | +0.00b | 6.55s | +744.00ms ⚠️ |
| dist/16.2c4c1846.js | 1.27kb | +0.00b | 5.81s | +507.00ms ⚠️ |
| dist/emoji.93ae61d3.js | 1.23kb | +0.00b | 6.03s | +602.00ms ⚠️ |
| dist/16.d17ab186.js | 1.23kb | +0.00b | 5.53s | +406.00ms ⚠️ |
| dist/16.436f097f.js | 1.21kb | +0.00b | 1.61s | +278.00ms ⚠️ |
| dist/16.b4da0acb.js | 1.19kb | +0.00b | 5.53s | +401.00ms ⚠️ |
| dist/16.02c68c4f.js | 1.18kb | +0.00b | 5.81s | +507.00ms ⚠️ |
| dist/heading5.e195aab1.js | 1.17kb | +0.00b | 6.55s | +744.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.2323af57.js | 1.15kb | +0.00b | 27.68s | +1.42s ⚠️ |
| dist/16.b46ef690.js | 1.14kb | +0.00b | 1.61s | +269.00ms ⚠️ |
| dist/16.4e3b0235.js | 1.14kb | +0.00b | 5.53s | +406.00ms ⚠️ |
| dist/16.9adb4ec7.js | 1.13kb | +0.00b | 5.53s | +406.00ms ⚠️ |
| dist/heading2.a2a1be3f.js | 1.11kb | +0.00b | 6.55s | +744.00ms ⚠️ |
| dist/16.3920766f.js | 1.10kb | +0.00b | 1.61s | +277.00ms ⚠️ |
| dist/mention.4d64e059.js | 1.07kb | +0.00b | 6.31s | +670.00ms ⚠️ |
| dist/heading4.ec980e93.js | 1.06kb | +0.00b | 6.55s | +744.00ms ⚠️ |
| dist/16.afcb09b6.js | 1.05kb | +0.00b | 5.81s | +507.00ms ⚠️ |
| dist/layout.52ba15f7.js | 1.05kb | +0.00b | 6.18s | +642.00ms ⚠️ |
| dist/divider.16defde6.js | 1.03kb | +0.00b | 6.03s | +602.00ms ⚠️ |
| dist/quote.aa25b83c.js | 1.02kb | +0.00b | 6.43s | +705.00ms ⚠️ |
| dist/action.9aa72b39.js | 1.00kb | +0.00b | 5.81s | +507.00ms ⚠️ |
| dist/decision.20244046.js | 1010.00b | +0.00b | 6.03s | +602.00ms ⚠️ |
| dist/panel-warning.97a0e3b3.js | 1008.00b | +0.00b | 6.31s | +671.00ms ⚠️ |
| dist/list.c9147a62.js | 978.00b | +0.00b | 6.18s | +642.00ms ⚠️ |
| dist/heading1.7d899ba5.js | 976.00b | +0.00b | 6.45s | +722.00ms ⚠️ |
| dist/panel-error.8d11e8b7.js | 906.00b | +0.00b | 6.31s | +670.00ms ⚠️ |
| dist/panel.a28c56ee.js | 897.00b | +0.00b | 6.43s | +705.00ms ⚠️ |
| dist/table.5a4874ee.js | 889.00b | +0.00b | 6.43s | +705.00ms ⚠️ |
| dist/panel-success.9b923e58.js | 845.00b | +0.00b | 6.31s | +671.00ms ⚠️ |
| dist/panel-note.97a4d541.js | 841.00b | +0.00b | 6.31s | +670.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.663369b9.js | 835.00b | +0.00b | 11.31s | -1.31m 🚀 |
| dist/media-viewer-analytics-error-boundary.0afa7031.js | 755.00b | +0.00b | 14.20s | +772.00ms ⚠️ |
| dist/index.html | 119.00b | +0.00b | 7.08s | +792.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.accd8b81.js | 2.72mb | -2.00b 🚀 | 267.00ms | -30.00ms 🚀 |
| dist/pdfRenderer.15a6a023.js | 1.11mb | +0.00b | 502.00ms | -36.00ms 🚀 |
| dist/editorView.e06544fb.js | 587.12kb | +0.00b | 551.00ms | -40.00ms 🚀 |
| dist/popup.6c453e3c.js | 169.96kb | +0.00b | 526.00ms | -29.00ms 🚀 |
| dist/Toolbar.83ac0b6e.js | 92.34kb | +0.00b | 551.00ms | -40.00ms 🚀 |
| dist/EmojiPickerComponent.ae601247.js | 75.29kb | +0.00b | 431.00ms | -28.00ms 🚀 |
| dist/media-viewer.7079364e.js | 72.03kb | +0.00b | 475.00ms | -35.00ms 🚀 |
| dist/card.645a5c17.js | 55.01kb | +0.00b | 474.00ms | -36.00ms 🚀 |
| dist/card.0313bac8.js | 52.78kb | +0.00b | 538.00ms | -37.00ms 🚀 |
| dist/Modal.19ffed32.js | 41.27kb | +0.00b | 281.00ms | -29.00ms 🚀 |
| dist/component.d7c29980.js | 30.70kb | +0.00b | 267.00ms | -29.00ms 🚀 |
| dist/DatePicker.0df05dac.js | 21.21kb | +0.00b | 430.00ms | -29.00ms 🚀 |
| dist/EmojiPickerComponent.7dc0ebd0.js | 18.23kb | +0.00b | 430.00ms | -30.00ms 🚀 |
| dist/dropzone.291ad844.js | 17.34kb | +0.00b | 538.00ms | -37.00ms 🚀 |
| dist/smartMediaEditor.125a19a7.js | 16.93kb | +0.00b | 550.00ms | -41.00ms 🚀 |
| dist/js.5733a017.js | 16.40kb | +0.00b | 281.00ms | -29.00ms 🚀 |
| dist/ui.fee77548.js | 13.97kb | +0.00b | 474.00ms | -36.00ms 🚀 |
| dist/workerHasher.1e706e77.js | 11.75kb | +0.00b | 454.00ms | -37.00ms 🚀 |
| dist/EmojiTypeAheadComponent.1990d771.js | 8.67kb | +0.00b | 454.00ms | -37.00ms 🚀 |
| dist/card.383bd9bd.js | 5.61kb | +0.00b | 475.00ms | -35.00ms 🚀 |
| dist/media-viewer.59c17dad.js | 3.92kb | +0.00b | 502.00ms | -36.00ms 🚀 |
| dist/EmojiPickerComponent.627aca6b.js | 3.58kb | +0.00b | 431.00ms | -29.00ms 🚀 |
| dist/index.ac562c80.css | 3.46kb | +0.00b | 585.00ms | -38.00ms 🚀 |
| dist/png-chunks-extract.a844fdb6.js | 3.41kb | +0.00b | 281.00ms | -29.00ms 🚀 |
| dist/dropzone.3c67f0a6.js | 3.23kb | +0.00b | 526.00ms | -30.00ms 🚀 |
| dist/Modal.0304b9ed.js | 3.08kb | +0.00b | 281.00ms | -29.00ms 🚀 |
| dist/clipboard.da112b6e.js | 2.75kb | +0.00b | 526.00ms | -30.00ms 🚀 |
| dist/16.290b4a43.js | 2.38kb | +0.00b | 281.00ms | -29.00ms 🚀 |
| dist/EmojiUploadComponent.d54a1fc6.js | 2.18kb | +0.00b | 454.00ms | -37.00ms 🚀 |
| dist/card.eea6ad8c.js | 2.03kb | +0.00b | 502.00ms | -37.00ms 🚀 |
| dist/ResourcedEmojiComponent.abe8596c.js | 2.00kb | +0.00b | 267.00ms | -30.00ms 🚀 |
| dist/date.231500d2.js | 1.73kb | +0.00b | 321.00ms | -28.00ms 🚀 |
| dist/16.d16f14ed.js | 1.73kb | +0.00b | 310.00ms | -34.00ms 🚀 |
| dist/images.80c92f5e.js | 1.67kb | +0.00b | 333.00ms | -20.00ms 🚀 |
| dist/feedback.d5590b34.js | 1.64kb | +0.00b | 409.00ms | -31.00ms 🚀 |
| dist/browser.75813db0.js | 1.61kb | +0.00b | 526.00ms | -30.00ms 🚀 |
| dist/workerHasher.6d5b22d7.js | 1.54kb | +0.00b | 454.00ms | -37.00ms 🚀 |
| dist/workerHasher.fa198233.js | 1.54kb | +0.00b | 539.00ms | -36.00ms 🚀 |
| dist/list-number.c35f9589.js | 1.46kb | +0.00b | 358.00ms | -25.00ms 🚀 |
| dist/status.8ab4226b.js | 1.46kb | +0.00b | 397.00ms | -30.00ms 🚀 |
| dist/16.9ccb5b28.js | 1.38kb | +0.00b | 307.00ms | -32.00ms 🚀 |
| dist/code.3dd71280.js | 1.38kb | +0.00b | 321.00ms | -28.00ms 🚀 |
| dist/16.c7751e34.js | 1.33kb | +0.00b | 307.00ms | -32.00ms 🚀 |
| dist/16.51049e2c.js | 1.32kb | +0.00b | 551.00ms | -40.00ms 🚀 |
| dist/16.18344f47.js | 1.31kb | +0.00b | 320.00ms | -29.00ms 🚀 |
| dist/link.2f4427e6.js | 1.30kb | +0.00b | 358.00ms | -25.00ms 🚀 |
| dist/heading6.fdbbdf4d.js | 1.30kb | +0.00b | 409.00ms | -30.00ms 🚀 |
| dist/heading3.2113b7a1.js | 1.29kb | +0.00b | 398.00ms | -30.00ms 🚀 |
| dist/16.2c4c1846.js | 1.27kb | +0.00b | 551.00ms | +207.00ms ⚠️ |
| dist/emoji.93ae61d3.js | 1.23kb | +0.00b | 333.00ms | -20.00ms 🚀 |
| dist/16.d17ab186.js | 1.23kb | +0.00b | 311.00ms | -33.00ms 🚀 |
| dist/16.436f097f.js | 1.21kb | +0.00b | 552.00ms | -39.00ms 🚀 |
| dist/16.b4da0acb.js | 1.19kb | +0.00b | 311.00ms | -33.00ms 🚀 |
| dist/16.02c68c4f.js | 1.18kb | +0.00b | 320.00ms | -29.00ms 🚀 |
| dist/heading5.e195aab1.js | 1.17kb | +0.00b | 409.00ms | -30.00ms 🚀 |
| dist/expand.450f3aea.js | 1.16kb | +0.00b | 409.00ms | -31.00ms 🚀 |
| dist/media-card-analytics-error-boundary.2323af57.js | 1.15kb | +0.00b | 502.00ms | -37.00ms 🚀 |
| dist/16.b46ef690.js | 1.14kb | +0.00b | 552.00ms | -39.00ms 🚀 |
| dist/16.4e3b0235.js | 1.14kb | +0.00b | 310.00ms | -281.00ms 🚀 |
| dist/16.9ba230ea.js | 1.13kb | +0.00b | 306.00ms | -33.00ms 🚀 |
| dist/16.a37df53b.js | 1.13kb | +0.00b | 307.00ms | -32.00ms 🚀 |
| dist/16.9adb4ec7.js | 1.13kb | +0.00b | 307.00ms | -32.00ms 🚀 |
| dist/heading2.a2a1be3f.js | 1.11kb | +0.00b | 398.00ms | -30.00ms 🚀 |
| dist/16.3920766f.js | 1.10kb | +0.00b | 552.00ms | -39.00ms 🚀 |
| dist/component.6600f83e.js | 1.07kb | +0.00b | 268.00ms | -28.00ms 🚀 |
| dist/mention.4d64e059.js | 1.07kb | +0.00b | 358.00ms | -26.00ms 🚀 |
| dist/heading4.ec980e93.js | 1.06kb | +0.00b | 407.00ms | -32.00ms 🚀 |
| dist/16.afcb09b6.js | 1.05kb | +0.00b | 311.00ms | -33.00ms 🚀 |
| dist/layout.52ba15f7.js | 1.05kb | +0.00b | 334.00ms | -19.00ms 🚀 |
| dist/divider.16defde6.js | 1.03kb | +0.00b | 333.00ms | -20.00ms 🚀 |
| dist/quote.aa25b83c.js | 1.02kb | +0.00b | 371.00ms | -33.00ms 🚀 |
| dist/action.9aa72b39.js | 1.00kb | +0.00b | 320.00ms | -29.00ms 🚀 |
| dist/decision.20244046.js | 1010.00b | +0.00b | 333.00ms | -20.00ms 🚀 |
| dist/panel-warning.97a0e3b3.js | 1008.00b | +0.00b | 371.00ms | -33.00ms 🚀 |
| dist/list.c9147a62.js | 978.00b | +0.00b | 358.00ms | -25.00ms 🚀 |
| dist/heading1.7d899ba5.js | 976.00b | +0.00b | 397.00ms | -30.00ms 🚀 |
| dist/panel-error.8d11e8b7.js | 906.00b | +0.00b | 358.00ms | -26.00ms 🚀 |
| dist/panel.a28c56ee.js | 897.00b | +0.00b | 371.00ms | -33.00ms 🚀 |
| dist/table.5a4874ee.js | 889.00b | +0.00b | 397.00ms | -30.00ms 🚀 |
| dist/panel-success.9b923e58.js | 845.00b | +0.00b | 370.00ms | -34.00ms 🚀 |
| dist/panel-note.97a4d541.js | 841.00b | +0.00b | 370.00ms | -33.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.f8ddfee3.js | 839.00b | +0.00b | 526.00ms | -30.00ms 🚀 |
| dist/media-card-analytics-error-boundary.663369b9.js | 835.00b | +0.00b | 474.00ms | -36.00ms 🚀 |
| dist/media-card-analytics-error-boundary.73ba4074.js | 835.00b | +0.00b | 539.00ms | -36.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.0afa7031.js | 755.00b | +0.00b | 502.00ms | -36.00ms 🚀 |
| dist/EmojiPickerComponent.2e54b9c6.js | 535.00b | +0.00b | 431.00ms | -29.00ms 🚀 |
| dist/simpleHasher.16c8bf41.js | 528.00b | +0.00b | 454.00ms | -37.00ms 🚀 |
| dist/simpleHasher.a29b2a8e.js | 528.00b | +0.00b | 539.00ms | -37.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 267.00ms | -30.00ms 🚀 |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/279504547722822157)**

File: packages/reporters/dev-server/src/HMRServer.js [hunk 1]
```diff
@@ -29,6 +34,8 @@ type HMRMessage =
       |},
     |};
 
+const FS_CONCURRENCY = 64;
```
Comment (mischnic) [inline packages/reporters/dev-server/src/HMRServer.js:37]: https://github.com/parcel-bundler/parcel/blob/069a2b1d773cb9fe02d621c2a281ecb47b5bf668/packages/shared/scope-hoisting/src/concat.js#L71

https://github.com/parcel-bundler/parcel/blob/069a2b1d773cb9fe02d621c2a281ecb47b5bf668/packages/packagers/js/src/JSPackager.js#L84

https://github.com/parcel-bundler/parcel/blob/069a2b1d773cb9fe02d621c2a281ecb47b5bf668/packages/packagers/css/src/CSSPackager.js#L21

Should we make this limit consistent? Is there a reason it's twice as large here?

Comment (wbinnssmith) [reply packages/reporters/dev-server/src/HMRServer.js:37]: One difference here is this is run in a reporter on a single thread, whereas packagers can run concurrently across threads, each with their own queue of open files.

**Review by mischnic — APPROVED** [review]

**Comment by mischnic** [comment]
> Should this happen in the Cache or even NodeFS instead?

A MemoryFS doesn't really need this limit, right?

**Comment by wbinnssmith** [comment]
> A MemoryFS doesn't really need this limit, right?

Yeah. What do you think of throttling at the NodeFS layer? @devongovett?

**Pull request merged**